### PR TITLE
Provide multicast isolation with in-container-netns iptables rules

### LINKED ISSUE #354: Multicast packets circumvent app isolation
With CoreOS out-of-the box:

```
core@core-01 ~ $ docker attach `sudo weave run 10.9.1.1/24 -ti bboreham/mcast:latest`
...
Status: Downloaded newer image for bboreham/mcast:latest
root@f7f69b63a0b7:/# 
root@f7f69b63a0b7:/# /home/mcreceive 224.2.2.4 5050
Received 3 bytes from 10.9.1.2: HI
Received 3 bytes from 10.9.2.2: HI
Received 3 bytes from 10.10.2.2: HI
Received 3 bytes from 192.168.2.2: HI
```

Now on the second machine:
```
core@core-02 ~ $ docker attach `sudo weave run 10.3.1.1/24 -ti bboreham/mcast:latest`
...
Status: Downloaded newer image for bboreham/mcast:latest
root@3fb6b80d2820:/# 
root@3fb6b80d2820:/# /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
HI
^C
root@3fb6b80d2820:/# exit
```
For some reason it didn't arrive on the first attemp, let's try again...
```
core@core-02 ~ $ docker attach `sudo weave run 10.9.1.2/24 -ti bboreham/mcast:latest`
root@711ca39b82d3:/# 
root@711ca39b82d3:/# echo HI | /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
root@711ca39b82d3:/# exit
core@core-02 ~ $ docker attach `sudo weave run 10.9.2.2/24 -ti bboreham/mcast:latest`
root@a44b9ab53823:/# echo HI | /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
root@a44b9ab53823:/# exit
core@core-02 ~ $ docker attach `sudo weave run 10.10.2.2/24 -ti bboreham/mcast:latest`
root@5e14e4888104:/# 
root@5e14e4888104:/# echo HI | /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
root@5e14e4888104:/# exit
core@core-02 ~ $ docker attach `sudo weave run 192.168.2.2/24 -ti bboreham/mcast:latest`
root@7f844a353553:/# 
root@7f844a353553:/# echo HI | /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
root@7f844a353553:/# 
```

To reproduce this, you can do:
```
git clone https://github.com/errordeveloper/weave-demos
cd weave-demos/felix; vagrant up
```
Then `vagrant ssh core-01` and `vagrant ssh core-02`.

----
Just to be sure, check `rp_fileter`:
```
$ for i in /proc/sys/net/ipv4/conf/*/rp_filter; do echo $i: `cat $i`; done
/proc/sys/net/ipv4/conf/all/rp_filter: 1
/proc/sys/net/ipv4/conf/default/rp_filter: 1
/proc/sys/net/ipv4/conf/docker0/rp_filter: 1
/proc/sys/net/ipv4/conf/dummy0/rp_filter: 1
/proc/sys/net/ipv4/conf/eth0/rp_filter: 1
/proc/sys/net/ipv4/conf/eth1/rp_filter: 1
/proc/sys/net/ipv4/conf/lo/rp_filter: 0
/proc/sys/net/ipv4/conf/veth0114561/rp_filter: 1
/proc/sys/net/ipv4/conf/veth2310286/rp_filter: 1
/proc/sys/net/ipv4/conf/veth3188b88/rp_filter: 1
/proc/sys/net/ipv4/conf/vethwepl1293/rp_filter: 1
/proc/sys/net/ipv4/conf/vethwepl1485/rp_filter: 1
/proc/sys/net/ipv4/conf/vethwepl1875/rp_filter: 1
/proc/sys/net/ipv4/conf/weave/rp_filter: 1
```
----
Let's try to run docker daemon with `--icc=false`...

If you use _weave-demos/felix_, do this
```
$ git checkout docker-icc-false
$ vagrant destroy -f ; vagrant up
```

and then back in.
```
core@core-01 ~ $ ps -ef | grep docker | grep icc
root       987     1  4 12:16 ?        00:00:12 docker --daemon --host=fd:// --insecure-registry=0.0.0.0/0 --icc=false
core@core-01 ~ $ docker attach `sudo weave run 10.9.1.1/24 -ti bboreham/mcast:latest`
...
Status: Downloaded newer image for bboreham/mcast:latest
root@3a11d59950e7:/# 
root@3a11d59950e7:/# /home/mcreceive 224.2.2.4 5050
Received 3 bytes from 10.9.1.2: HI
Received 3 bytes from 10.9.5.2: HI
Received 3 bytes from 192.168.5.2: HI

^C
root@3a11d59950e7:/# exit
```

I have repeated most of the same commands on `core-02`:
```
core@core-02 ~ $ docker attach `sudo weave run 10.9.1.2/24 -ti bboreham/mcast:latest`
root@2a43aaed4e03:/# 
root@2a43aaed4e03:/# echo HI | /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
root@2a43aaed4e03:/# exit
core@core-02 ~ $ docker attach `sudo weave run 10.9.5.2/24 -ti bboreham/mcast:latest`
root@ce66b715f7de:/# 
root@ce66b715f7de:/# echo HI | /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
root@ce66b715f7de:/# exit
core@core-02 ~ $ docker attach `sudo weave run 192.168.5.2/24 -ti bboreham/mcast:latest`
root@bca0694a2f9b:/# 
root@bca0694a2f9b:/# echo HI | /home/mcsend 224.2.2.4 5050
Begin typing (return to send, ctrl-C to quit):
root@bca0694a2f9b:/# exit
```

### PR DESCRIPTION
**Pull request by awh**
Fixes #354.

### CODE REVIEW

**Comment by awh** [comment]
I've only tested this cursorily; looking for initial feedback.

File: net/veth.go [hunk 1]
```diff
@@ -194,11 +224,48 @@ func DetachContainer(ns netns.NsHandle, id, ifName string, cidrs []*net.IPNet) e
...
+		subnets := subnets(addrs)
+		rules, err := ipt.List("filter", "INPUT")
+		if err != nil {
+			return err
+		}
+		for _, rule := range rules {
+			ps := strings.Split(rule, " ")
+			if len(ps) == 10 &&
+				ps[0] == "-A" && ps[2] == "-s" && ps[4] == "-d" && ps[5] == "224.0.0.0/4" &&
+				ps[6] == "-i" && ps[7] == ifName && ps[8] == "-j" && ps[9] == "ACCEPT" {
```
Comment (rade) [inline net/veth.go]: The horror... would a regexp be more readable?

Comment (awh) [inline net/veth.go]: I considered it, but it seemed overkill

**Comment by awh** [comment]
I have tested this manually - ready to merge IMO.

**Comment by awh** [comment]
Can you look at this @rade? - it's ready to merge IMO - Martynas is busy with the netns stuff atm...

**Pull request merged**